### PR TITLE
feat(Clarity2): Comparisons for string and buffer types

### LINKED ISSUE #2618: Extension contract: buff, string-ascii, and string-utf8 comparison
There needs to be an extension contract that makes all sequence-types comparable (`<`, `<=`, `==`, `!=`, `>`, `>=`).  It could take arguments that have the type that represents the maximum allowed of the sequences, and has runtime equal to the longest of the two compared sequences.

### PR DESCRIPTION
**Pull request by gregorycoppola**
Fixes #2618

This PR adds the ability to do the following comparisons:
* `<`, `>`, `<=`, `>=`
for the following additional types:
* `string-ascii
* string-utf8
* buff

The ability to compare these additional types was already present for '=='.

The new behaviors are gated on the version being Clarity2.

### CODE REVIEW

File: src/vm/analysis/type_checker/mod.rs
Comment (jcnelson) [inline src/vm/analysis/type_checker/mod.rs:234]: Given the number of matching criteria, I wonder if it makes more sense to introduce a `fn is_comparable(clarity_version: ClarityVersion) -> bool` method in `Value` that simply tells you whether or not arithmetic comparison operators can apply to `Value`.  Then, you'd only need to check that `first` and `second` are comparable in this version of Clarity, and then check that they have the same type.

File: src/vm/analysis/type_checker/mod.rs
Comment (jcnelson) [inline src/vm/analysis/type_checker/mod.rs:562]: Can you just derive `Copy` for `ClarityVersion`?

File: src/vm/tests/simple_apply_eval.rs
Comment (jcnelson) [inline src/vm/tests/simple_apply_eval.rs:637]: Can you expand test coverage to include strings and buffers of _different_ lengths?  The `PartialOrd` trait is automatically derived for `Vec<T>` if it is implemented for `T`, so since `T` is either `u8` or `Vec<u8>` in this case, we should expect the following to be true:

`(< "" "a")`
`(> "a" "")`
`(<= "" "a")`
`(>= u"a" u"")`
`(< u"" u"a")`
`(> u"a" u"")`
`(<= u"" u"a")`
`(>= u"a" u"")`
`(< 0x 0x01)`
`(> 0x01 0x)`
`(<= 0x 0x01)`
`(>= 0x01 0x)`

Comment (gregorycoppola) [reply src/vm/tests/simple_apply_eval.rs]: I updated the tests, to 1) use different lengths, and 2) make the set of tests comprehensive.

**Review by gregorycoppola — COMMENTED** [review]
thanks for the careful comments

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:380]: More of a question for the group -- is it `Clarity1` and `Clarity2`, or is it `Clarity 1` and `Clarity 2`, or is it something else?

File: src/vm/docs/mod.rs
Comment (jcnelson) [inline src/vm/docs/mod.rs:385]: Do you want to add examples with buffers and strings of unequal lengths?  Or, maybe state that the comparison is a lexicographical comparison?

File: src/vm/functions/arithmetic.rs
Comment (jcnelson) [inline src/vm/functions/arithmetic.rs:101]: is 2 or greater

File: src/vm/types/signatures.rs
Comment (jcnelson) [inline src/vm/types/signatures.rs:640]: These will be removed in favor of the code here, right? https://github.com/blockstack/stacks-blockchain/pull/2721#discussion_r662449453

**Review by jcnelson — APPROVED** [review]
This LGTM; just had a few minor comments to address before merging.  Thanks @gregorycoppola!

File: src/vm/types/signatures.rs
Comment (gregorycoppola) [reply src/vm/types/signatures.rs:640]: yes

**Review by gregorycoppola — COMMENTED** [review]
Thanks for the review, Jude.. I added @pavitthrap for second review.

File: src/vm/tests/mod.rs
Comment (pavitthrap) [inline src/vm/tests/mod.rs:106]: function seems redundant

File: src/vm/tests/simple_apply_eval.rs
Comment (pavitthrap) [inline src/vm/tests/simple_apply_eval.rs:552]: for both the success and error tests, could add tests where the sequences are of different lengths

File: src/vm/tests/simple_apply_eval.rs
Comment (pavitthrap) [inline src/vm/tests/simple_apply_eval.rs:672]: should evaluate these tests in clarity2 as well

**Review by pavitthrap — APPROVED** [review]
lgtm, just left some feedback about testing!

File: src/vm/tests/mod.rs
Comment (gregorycoppola) [reply src/vm/tests/mod.rs]: ok, good point.. removed it

File: src/vm/tests/simple_apply_eval.rs
Comment (gregorycoppola) [reply src/vm/tests/simple_apply_eval.rs:552]: I changed teh success tests from this section to use different lengths.. note that the tests in `test_sequence_comparisons_clarity2` also use different lengths.

File: src/vm/tests/simple_apply_eval.rs
Comment (gregorycoppola) [reply src/vm/tests/simple_apply_eval.rs:675]: ok.. good idea.

File: src/vm/docs/mod.rs
Comment (gregorycoppola) [reply src/vm/docs/mod.rs:444]: ok sure

**Review by gregorycoppola — COMMENTED** [review]
Thanks for the reviews, everyone.

**Pull request merged**
